Automate test scenario for job array with ASAP and with standing rese…

Klovely opened the pull request:
…rvation

Automate test scenario : 
- job array converted to ASAP reservation
- job array within standing reservation

#### Describe Your Change
Added a test to cover below scenario
- job array converted to ASAP reservation
- job array within standing reservation

#### Attach Test and Valgrind Logs/Output
[Execution_logs_myresult4.txt](https://github.com/PBSPro/pbspro/files/3349697/Execution_logs_myresult4.txt)

dinesh1306 commented:
Looks Good.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+        j = Job(TEST_USER, attrs=a)
+        j.set_sleep_time(20)
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, jid)
```
why do we need to submit this job?

Klovely replied:
Submitted this job to take up all resources , before submitting job-array .

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+            attrop=PTL_AND, id=subjid[2])
+        self.server.expect(
+            JOB, {'job_state': 'Q', 'queue': rid1_q},
+            attrop=PTL_AND, id=subjid[8])
```
why the '8'th subjob is being checked?

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+
+        # Check for finished jobs by issuing the command qstat
+        self.server.expect(JOB, {'job_state': 'F', 'substate': '93'},
+                           extend='x', id=subjid[3], offset=65)
```
Why is offset set to 65? since this is the 4th job, I assume this job would have ended with the previous reservation instance so ideally should be finished by now.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+                           id=jid, extend='x')
+
+    @skipOnCpuSet
+    def test_job_array_within_recurring_reservation(self):
```
The name is confusing because the recurring reservation is standing reservation and then it has the same name as the previous test.

saritakh commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+        # Wait for job J1 to finish
+        self.logger.info('Waiting for job J1 to be finished')
+        self.server.expect(JOB, {'job_state': 'F'},
+                           extend='x', id=jid, offset=20)
```
I think this wait for job1 to finish is needed after the ASAP reservation submission.

saritakh commented:
@Klovely I have a generic comment here: I suggest to add relevant tags to all the new test cases as given in [PTL test guidelines](https://pbspro.atlassian.net/wiki/spaces/DG/pages/519372968/Guidelines+for+Writing+PTL+Tests). This would help a lot in future for test categorization.
In my understanding the tag that is applicable to the test suite of this PR is 'reservations'. I suggest to specify tag all at once at suite level as @tags('reservations').

Klovely commented:
Testing logs

[Execution_logs_latest_after_change.txt](https://github.com/PBSPro/pbspro/files/3386071/Execution_logs_latest_after_change.txt)

Klovely commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1123,409 @@ def test_reservations_with_expired_subjobs(self):
+        # Wait for standing reservation first instance to start
+        exp_attr = {'reserve_state': (MATCH_RE, "RESV_RUNNING|5")}
+        self.server.expect(RESV, exp_attr, id=rid, offset=10)
+        self.server.expect(RESV, {'reserve_index': 1}, id=rid)
```
Just to confirm that first instance is running

Klovely commented:
@prakashcv13 , @arungrover and @saritakh : I have addressed your review comments in latest commit  and attached latest test execution logs. 
Please review the latest commit

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1327,53 +1328,47 @@ def test_standing_resv_with_job_array(self):
         self.logger.info(
-            'Waiting 60 seconds for first instance of reservation to Finish')
-        time.sleep(60)
+            'Waiting 20 seconds for first instance of reservation to Finish')
```
why are you waiting for 20 seconds here? The last reservation was only 10 seconds long. So Ideally after 10 seconds it should get back to RESV_CONFIRMED state again.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1327,53 +1328,47 @@ def test_standing_resv_with_job_array(self):
+            'Waiting 50 sec for second  instance of reservation to start')
         exp_attr = {'reserve_state': (MATCH_RE, "RESV_RUNNING|5")}
-        self.server.expect(RESV, exp_attr, id=rid)
+        self.server.expect(RESV, exp_attr, id=rid, offset=50, interval=10)
```
you already have an offset of 50 seconds, you do not need an interval of 10

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1327,53 +1328,47 @@ def test_standing_resv_with_job_array(self):
+        self.server.expect(RESV, {'reserve_index': 2}, id=rid)
         # Wait for reservations to be finished
         msg = "Que;" + rid_q + ";deleted at request of pbs_server@"
         self.server.log_match(msg, starttime=end, interval=20)
```
I'd suggest you reduce the interval to a second because the reservation is already running and it should get deleted  10 seconds from now.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1244,60 +1254,51 @@ def test_ASAP_resv_with_job_array(self):
         # Check status of the sub-job array using qstat -fx once reservation
         # ends
-        self.server.expect(JOB, {'job_state': 'F', 'substate': '92',
+        self.server.expect(JOB, {'job_state': 'F',
```
any reason why you removed check for substate here?

Klovely replied:
Reason: I found a issue while checking substate ,there is a bug in job array I m going to file separate issue for that, job-array is showing substate 92 only when all subjobs finished within reservation with exit status 0, if any of sub job terminate with exit status not equal to zero then it mark substate as 91 for those subjob also which it complete with exit status 0.

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1487,45 +1475,58 @@ def test_job_array_within_recurring_reservation(self):
-            self.server.expect(JOB, {'job_state': 'F', 'substate': '93'},
-                               extend='x', id=subjob)
+            'Waiting 30 sec job-array 1 and 2 to be finished')
+        self.server.expect(JOB, {'job_state=F': 3}, extend='xt',
```
You have already checked that reservation is running, you can reduce the offset to 20 or less

arungrover commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1487,45 +1475,58 @@ def test_job_array_within_recurring_reservation(self):
-        # Wait for reservations to be finished
+        # Wait for reservation to be finished
         msg = "Que;" + rid_q + ";deleted at request of pbs_server@"
         self.server.log_match(msg, starttime=end, interval=20)
```
Here too, consider reducing the interval to a second because it will not take very long to end the reservation.

Klovely commented:
Testing logs :- 
[Execution_logs_after_latest_change.txt](https://github.com/PBSPro/pbspro/files/3390798/Execution_logs_after_latest_change.txt)

Klovely commented:
Thanks @arungrover  for your review comments on my last commit.
I have addressed your comments in latest commit and attached latest test execution logs.
Please review the latest commit

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+
+        # Check for finished jobs by issuing the command qstat
+        # all sub-jobs should finished
+        # in reservation it has substate as 92
```
this comment, does not explain why you expect substate to be 92.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+                           extend='xt', id=jid)
+
+        # Submit a standing reservation, that gets confirmed.  Submission time
+        # is HHMM
```
we can remove this comment.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+        for i in range(1, 6):
+            subjid.append(j.create_subjob_id(jid, i))
+        # Wait for standing reservation first instance to start
+        # Verify one sub-job should running and others in queued
```
Verify one subjob starts running and others remain queued.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+        self.server.sigjob(jobid=subjid[0], signal="suspend")
+        self.server.expect(JOB, {'job_state': 'S'}, id=subjid[0])
+        self.server.expect(JOB, {'job_state': 'R'}, id=subjid[1])
+        # Resume suspend sub-job[1] and verify sub-job[1] should
```
remove the word "suspend"

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+        self.server.expect(JOB, {'job_state': 'S'}, id=subjid[0])
+        self.server.expect(JOB, {'job_state': 'R'}, id=subjid[1])
+        # Resume suspend sub-job[1] and verify sub-job[1] should
+        # run once resource available
```
once resources are available.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+        self.server.expect(RESV, exp_attr, id=rid)
+        rid_q = rid.split(".")[0]
+        # Submit 3 job-array within reservation with sleep time longer
+        # than instance reservation
```
than "instance duration"

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+                                 'queue': rid_q}, extend='x',
+                           attrop=PTL_AND, id=subjid[5])
+
+        # Check for  sub-jobs status of job-array 1 and 2
```
2 spaces before sub-jobs.

prakashcv13 commented on test/tests/functional/pbs_reservations.py:
```diff
@@ -1123,3 +1124,407 @@ def test_reservations_with_expired_subjobs(self):
+        # Check for  sub-jobs status of job-array 1 and 2
+        # as all sub-jobs from job-array 1 and 2 exceed walltime
+        # and failed to complete within an instance of reservation
+        # so it should substate as 93
```
exceed walltime of the reservation, so they will not complete running, so the substate of these subjobs should be 93.

Klovely commented:
Latest testing log :-
          
[Execution_logs2_latest_after_change.txt](https://github.com/PBSPro/pbspro/files/3392928/Execution_logs2_latest_after_change.txt)

Klovely commented:
Thanks @prakashcv13 for your review comments on my last commit.
 I have addressed your comments .
Please review the latest commit
